xtra-libp2p/src/dialer.rs

```diff
@@ -1,107 +1,124 @@
-
-        connected_peers
-            .contains(
-                &self
-                    .connect_address
-                    .clone()
-                    .extract_peer_id()
-                    .ok_or(Error::InvalidPeerId)?,
-            )
-            .then(|| ())
-            .ok_or(Error::ConnectionDropped)
-    }
-
     async fn connect(&self) -> Result<(), Error> {
         self.endpoint
             .send(Connect(self.connect_address.clone()))
             .await
             .map_err(|_| Error::NoEndpoint)?
             .map_err(|e| Error::Failed { source: anyhow!(e) })
     }
+
+    fn stop_with_error(&mut self, e: Error, ctx: &mut xtra::Context<Self>) {
+        self.stop_reason = Some(e);
+        ctx.stop();
+    }
 }
 
 #[async_trait]
 impl xtra::Actor for Actor {
     type Stop = Error;
 
     async fn started(&mut self, ctx: &mut xtra::Context<Self>) {
-        let this = ctx.address().expect("we are alive");
+        match self
+            .connect_address
+            .clone()
+            .extract_peer_id()
+            .ok_or(Error::InvalidPeerId)
+        {
+            Ok(peer_id) => self.peer_id = Some(peer_id),
+            Err(e) => {
+                self.stop_with_error(e, ctx);
+            }
+        }
 
         if let Err(e) = self.connect().await {
-            self.stop_reason = Some(e);
-            ctx.stop();
+            self.stop_with_error(e, ctx);
         }
 
-        // Only start checking the connection after it had enough time to be established
-        tokio::time::sleep(CHECK_INTERVAL).await;
-        self.tasks
-            .add(this.send_interval(CHECK_INTERVAL, || CheckConnection));
+        // TODO: add connection logic
+        // tokio::time::sleep(CONNECTION_TIMEOUT).await;
+        //
+        // if !self.connected {
+        //     self.stop_with_error(
+        //         Error::Failed {
+        //             source: anyhow!("Connection timeout lapsed"),
+        //         },
+        //         ctx,
+        //     );
+        // }
     }
 
     async fn stopped(self) -> Self::Stop {
         self.stop_reason.unwrap_or(Error::Unspecified)
     }
 }
 
+impl Actor {
+    fn peer_id(&self) -> PeerId {
+        self.peer_id
+            .expect("to always have peer id if successfully started")
+    }
+}
+
 #[xtra_productivity]
 impl Actor {
     async fn handle(&mut self, msg: Error, ctx: &mut xtra::Context<Self>) {
-        self.stop_reason = Some(msg);
-        ctx.stop();
+        self.stop_with_error(msg, ctx);
     }
+}
 
-    async fn handle(&mut self, _msg: CheckConnection, ctx: &mut xtra::Context<Self>) {
-        if let Err(e) = self.check_connection_active_in_endpoint().await {
-            self.stop_reason = Some(e);
-            ctx.stop();
+#[xtra_productivity(message_impl = false)]
+impl Actor {
+    async fn handle(&mut self, msg: connection_monitor::ConnectionsEstablished) {
+        if msg.peers.contains(&self.peer_id()) {
+            self.connected = true;
```

Comment:
the thinking behind was as such:
we should know whether the initial connection has succeeded within the specified timeout - otherwise we might be hanging with a connection attempt.... (this is to cover some edge cases where we could be left hanging, and just trigger a reconnect instead)